### ISSUE TITLE
uinput-based backend for Linux

### ISSUE THREAD

**Issue by TMiguelT** [moses-palmer/pynput#184]
Hi, I've noticed that the only support `pynput` has for Linux is via X11, which obviously won't work in Wayland environments. It looks like we should be using a `uinput`-based solution, which is a kernel module that works on any graphics server. And fortunately, [python-evdev](https://github.com/gvalkov/python-evdev) provides a nice wrapper for its use.

Is there any interest in a PR to add `uinput` support? Because `uinput` requires elevated permissions, this won't always replace the X11 option, so Linux users would need to choose between X11 and `uinput` somehow. Is this feasible?

**Comment by moses-palmer** [comment]
I have certainly considered adding this feature---I even had a local test branch at one point.

The major problem I see is mapping keyboard events to characters---we would need a layout description. As a matter of fact, I also started writing a Python parser for X keyboard layout files... Do you have a good solution to this problem?

**Comment by TMiguelT** [comment]
Wow, yep this is a tricky area. X and Wayland both use the xkb format, but Wayland doesn't provide an easy way to access the current map, and I believe `setxkbmap -query` is only designed to work under X11. Gnome _does_ provide a keymap, so `gsettings get org.gnome.desktop.input-sources mru-sources` will work, but that's obviously only going to work under Gnome.

However, I think I've had a better idea. While the Wayland protocol doesn't specify how the keymap loaded, it *does* require that the server provides the keymap to Wayland clients. So we could momentarily pretend to be a direct Wayland client (using something like [pywayland](https://pywayland.readthedocs.io/en/latest), listen for the `keymap` event, and then close the client. Apparently, this event provides a "a file descriptor to the client which can be memory- mapped to provide a keyboard mapping description", which I think would be sufficient for translating the inputs. 

Now obviously this would only work for Wayland, but that provides a nice decision point for `pynput`. If we're on Linux, we check if the session is Wayland or X11. If Wayland, we attempt the process above, and send keys using `uinput`. If X11, we keep the current code, which is X-specific.

Thoughts?

**Comment by TMiguelT** [comment]
Actually it might be better to keep the Wayland client open during the lifetime of a `Controller`, so that it can listen for new keymap events and update accordingly.

**Comment by TMiguelT** [comment]
Another thought. Apparently there's a lower-level keymap concept in Linux, described here: https://linux.die.net/man/5/keymaps. This keymap approach seems to be the same one used by [uinputchars](https://github.com/piater/uinputchars), which works excellently for me, even with my non-standard layout. Perhaps it would be easier to load and parse the Linux keymap via `dumpkeys` (or maybe a C API if there is one), and then feed key presses into `uinput` via `evdev` and the keymap. This should mean we rely on nothing except for the Kernel, which is even better than my Wayland suggestion proposed above.

**Comment by TMiguelT** [comment]
However, accessing the keymap via `dumpkeys` requires sudo, so that's a downside of this approach.

**Comment by moses-palmer** [comment]
Thank you very much for your thorough research!

The requirement on superuser privileges for `dumpkeys` should not be a problem, since this approach requires that to begin with to use `uinput`, right? Granted, a solution that does not depend on an external binary would be preferable, I think. I'm sure the source of `dumpkeys` would reveal how.

**Comment by TMiguelT** [comment]
That's true, but you can do things like `chmod`ing `/dev/uinput`, or adding it to a group that you're part of, so that you can then write to it without needing sudo anymore. Whereas I'm not sure if you can relax the permissions for `dumpkeys`.

**Comment by TMiguelT** [comment]
> a solution that does not depend on an external binary would be preferable, I think

True. The other option is using the C API, apparently [like this](https://github.com/piater/uinputchars/blob/8491b2c50e9c46847e65b1ef365eaf595241341f/keymap.c#L106-L139)

**Comment by moses-palmer** [comment]
Again, thank you very much for your research.

I too have concluded that *Wayland* integration may be possible in the future, but as you note, integration is not yet possible, at least not without integrating rather deeply with the various desktop environments. Contrary to you, however, I think the `uintput` route would make sense as an alternative, since it would work even without a graphical environment. One can only dream...

**Comment by TMiguelT** [comment]
Oh no, I still think a low-level kernel-based input would be useful. I'm just not sure I can justify doing it myself at the moment.

**Comment by cjbassi** [comment]
What about using ydotool:

https://github.com/ReimuNotMoe/ydotool

It's a much more naive approach but it should work, at least for the time being.

**Comment by TMiguelT** [comment]
I think it would be better to use a python module that can be installed as a simple pip dependency.

**Comment by cjbassi** [comment]
I agree, but I'm wondering if this would be a good temporary solution while Wayland figures itself out.

**Comment by cjbassi** [comment]
Yeah actually a better solution I think would be to go through uinput ourselves. I've already got an implementation for this in Python here:

https://github.com/osprey-voice/osprey/tree/master/osprey

It's a little spread out between the voice.py, evdev.py, and keys.py files. The main issues are that the question mark key currently does not work, and this requires running the program either as sudo or changing permissions on dev/uinput.

I can open a PR for adding this, would that be all right?

**Comment by TMiguelT** [comment]
Yes, directly interfacing with `uinput` makes sense. But why not use, for example [python-evdev](https://github.com/gvalkov/python-evdev)?

From previous discussion I think this is something that would get merged. Incidentally if you need this now, I think [keyboard](https://github.com/boppreh/keyboard) already provides a uinput interface.

**Comment by moses-palmer** [comment]
Inspired by this discussion, I have started work on an *evdev* based backend.

The issues I see initially are:
* I assume that the possible characters typable will be limited by the current keyboard layout.
* I think that certain parts of the mouse interface will be difficult to implement, as I doubt there is a concept of absolute positions at this level.

**Comment by dp0s** [comment]
Hello,
I have actually written a working solution, which is combining the abilities of pynput and evdev into a single interface because that's what I need every day. It works stable on my machine, but I am hesitant to publish it as I have no time for maintainance and not even to write a documentation right now. But if you are working on it anyway, you might take inspiration from my code and save work. 

I solved 2 major problems with Python-evdev:
1. You need to manually check which devices of which type are available and update the available devices regularly. That is a bit tricky, so I wrote a special EvdevDeviceUpdater class which can be reused in other projects.

2. Translating the keyboard layout is in fact necessary. I use the xkb-information (saved  as easily readable json). The user then needs to manually specify which translation is required.

About the mouse issue:
Python-evdev collects whatever the hardware is sending. My touchpad is sending the absolute x-y-coordinates, where the user is clicking inside the touchpad field and it does not care about what that actually does on the screen. So without from the x-server, python evdev will almost no chance to provide the same functionality as pynput does at the moment. 

As soon as I find the time, I can upload my project code and put the link here if you like.

**Comment by dp0s** [comment]
I just remembered a problem - my code is crashing with the latest release of python-evdev, I am still running an older version.. So maybe I need to change my code again before it can be used properly.

**Comment by dp0s** [comment]
Hello again, I published my project which includes the "evdev_preperation" package:

https://github.com/dp0s/Dpowers/tree/master/Dlib/Dhelpers/evdev_preperation

It is based on python-evdev 0.7 and works without async. If you want to use it in your projects, I could update it to the latest evdev version and publish it as seperate package. 

There are three modules that might be of interest:
1. A uinput wrapper that allows to use a single uinput instance while the backing capabilities can be changed after creation.
2. An adhanced InputEvent class (including categorization) and an input collector based on it.
3. A DeviceUpdater that is checking regularly if devices have been added/removed.

Please tell me if you find them useful.

As examples for the usage, you can have a look at:

https://github.com/dp0s/Dpowers/blob/master/Dlib/Dpowers/events/keybpower/adapt_evdev.py

https://github.com/dp0s/Dpowers/blob/master/Dlib/Dpowers/events/keybpower/adapt_evdev.py

The latter is a bit more complex because I created a larger framework that can be coupled to different backends (also to pynput). It allows blocking/remapping of single keys with evdev. I am not sure if it is stable on other computers than mine though. If you like to test it, let me know.

**Comment by dp0s** [comment]
About the problem of the Keyboard Layouts, have a look at the following package:

https://github.com/dp0s/Dpowers/tree/master/Dlib/Dhelpers/KeyboardLayouts

It includes a folder containing common xkb layouts as json.
Example of useage:

from Delpers.KeyboardLayouts import Layout
translation_dict = Layout("us") > Layout("de")

This dictionary allows you to manually translate the key names received from evdev to get the corresponding German keys (in my case).
The json files also contain infos about modified key names of each layout for a number of modifier keys. With this one could manually find out the actual modified output sent to the system even if evdev doesn't know about it. I don't need it so far.

**Comment by dp0s** [comment]
Finally I want to say a big thanks to Moses Palmér for creating pynput because it is the most important backend to my framework and helps me everyday. :-) 
Also thanks for creating pystray, I am using it as a backend too.

**Comment by moses-palmer** [comment]
Thank you for your kind words :-)

I did have a look at your project, and it provided helpful input. Thank you!

Here is a quick progress report:

At the moment I have a working keyboard controller with the following caveats:
1. At the moment the virtual key mapping is read using `dumpkeys` as suggested by @cjbassi. In theory, the `dumpkeys` code could be ported---I implemented as far at acquiring a working handle to the console and reading the keyboard type---but the remaining functionality would require quite a bit of additional work.
2. The mapped keys thus correspond to what is available in a VT, which is a quite limited subset of what you will find mapped under X/Wayland. Since we sent low-level keyboard events, those other letters would probably be rendered, but *pynput* has no way of translating exotic characters to the required modifier/letter key sequences.
3. My hope that this would prove to be a viable solution under *Wayland* was unfortunately unfounded: when running the *uinput* based backed under X or in a virtual terminal without a graphic server, it works, but the modifier state is very inconsistent under Wayland, so capital letters and those requiring <kbd>AltGr</kbd> do not work as expected.
